cabal-install/Distribution/Client/ProjectPlanning/Types.hs

```diff
@@ -91,16 +107,155 @@ data ElaboratedSharedConfig
+-- 'ElaboratedConfiguredPackage' which indicates that the "package"
+-- in question is actually a single component to be built.  Arguably
+-- it would be clearer if there were an ADT which branched into
+-- package work items and component work items, but I've structured
+-- it this way to minimize change to the existing code (which I
+-- don't feel qualified to rewrite.)
+data ElaboratedComponent
+   = ElaboratedComponent {
+    -- | The munged package identifier of this component.
+    -- Invariant that the 'Version' of this 'PackageId'
+    -- matches that in 'ElaboratedPackage'.
+    elabComponentPackageId :: PackageId,
+    -- | The name of the component to be built
+    elabComponent :: CD.Component,
+    -- | The name of the component to be built.  Nothing if
+    -- it's a setup dep.
+    elabComponentName :: Maybe ComponentName,
+    -- | The ID of the component to be built
+    elabComponentId :: UnitId,
+    -- | Dependencies of this component
+    elabComponentDependencies :: [ConfiguredId],
+    -- | The order-only dependencies of this component; e.g.,
+    -- if you depend on an executable it goes here.
+    elabComponentExeDependencies :: [UnitId],
+    -- | The 'ElaboratedPackage' this component came from
+    elabComponentPackage :: ElaboratedPackage,
+    -- | What in this component should we build (TRANSIENT, see 'pkgBuildTargets')
+    elabComponentBuildTargets :: [SubComponentTarget],
+    -- | Should we REPL this component (TRANSIENT, see 'pkgReplTarget')
+    elabComponentReplTarget :: Maybe SubComponentTarget,
+    -- | Should we Haddock this component (TRANSIENT, see 'pkgBuildHaddocks')
+    elabComponentBuildHaddocks :: Bool
+    -- NB: Careful, if you add elabComponentInstallDirs, need
+    -- to adjust 'packageHashInputs'!!!
+   }
+  deriving (Eq, Show, Generic)
+
+instance Binary ElaboratedComponent
+
+instance Package ElaboratedComponent where
+    -- NB: DON'T return the munged ID by default.
+    -- This is needed in projAllPkgs in
+    -- Distribution.Client.ProjectOrchestration.  A bit
+    -- too finely balanced for my liking.
+    packageId = packageId . elabComponentPackage
+
+instance HasConfiguredId ElaboratedComponent where
+    -- TODO: Munged or not? Who knows!
+    configuredId comp = ConfiguredId (packageId comp) (installedUnitId comp)
+
+instance HasUnitId ElaboratedComponent where
+    installedUnitId = elabComponentId
+
+instance IsNode ElaboratedComponent where
+    type Key ElaboratedComponent = UnitId
+    nodeKey = elabComponentId
+    nodeNeighbors comp =
+           map installedUnitId (elabComponentDependencies comp)
+        ++ elabComponentExeDependencies comp
+
+data ElaboratedPackage
+   = ElaboratedPackage {
 
        pkgInstalledId :: InstalledPackageId,
        pkgSourceId    :: PackageId,
 
-       -- | TODO: [code cleanup] we don't need this, just a few bits from it:
-       --   build type, spec version
        pkgDescription :: Cabal.PackageDescription,
```

Comment:
Oh, I preemptively removed this comment because Backpack needs a pile of things from `pkgDescription`.